_posts/2016-10-21-unit-test-e1-async-test.md

```diff
@@ -0,0 +1,146 @@
+---
+layout: post
+title: Unit test tips & tricks - Episode 1 - Avoid timeout in async test
+tip-number: 08
+tip-username: Sang D. Ngo
+tip-username-profile: https://github.com/heroandtn3
+tip-tldr: How to avoid timeout error in async test cases
+---
+
+We write unit test everywhere in OpenPaas, some of them are async, let's see an example:
+
+```javascript
+getModule()
+  .findByDomainId()
+  .then(function(configuration) {
+    expect(configuration.modules).to.have.length(2);
+    done();
+  });
+```
+
+It won't be a problem if it passes, but what if it fails?
+
+```
+Error: timeout of 20000ms exceeded. Ensure the done() callback is being called in this test.
+```
+
+Well, the test case got hang for 20s then printed out an error message that isn'
+helpful at all. Why the `done()` was not called? Did the promise reject? Did
+the promise take too much time to resolve? You start debuging the failed test by
+commenting the `expect` statement:
+
+```javascript
+getModule()
+  .findByDomainId()
+  .then(function(configuration) {
+    //expect(configuration.modules).to.have.length(2);
+    done();
+  });
+```
+
+And it passed!
+
+```
+Done, without errors.
+```
+
+Obviously, the problem comes from the `expect` statement, you continue with:
+
+```javascript
+getModule()
+  .findByDomainId()
+  .then(function(configuration) {
+    console.log('Length:', configuration.modules.length);
+    //expect(configuration.modules).to.have.length(2);
+    done();
+  });
+```
+
+```
+Length: 1
+```
+
+Cool! Finally, you found the source of the problem. You fix your code or your mock,
+the test comes green, you feel so happy.
+
+Well, it would not be easy like that when your test fails on Jenkins,
+especially with random failure. You may already knew, sometimes we have random failure
+on Jenkins and the error message is just `Error: timeout of 20000ms exceeded`,
```

Comment:
The sentence is toooooooo long.
You should replace 'jenkins' with a more generic term